common/model/cpptransaction.cc

```diff
@@ -0,0 +1,247 @@
+ */
+
+#include "common/model/cpptransaction.h"
+#include "common/trinary/trit_array.h"
+
+namespace iota {
+namespace model {
+Transaction::Transaction() { iota_transaction = transaction_new(); }
+
+Transaction::Transaction(const std::string &trytes) {
+  iota_transaction = transaction_new();
+  size_t to_len = trytes.size() * 3;
+  size_t flex_len = flex_trits_num_for_trits(to_len);
+  flex_trit_t trits[flex_len];
+  tryte_to_flex_trit(trits, to_len, (const tryte_t *)trytes.data(),
+                     trytes.size(), trytes.size());
+  transaction_deserialize_from_trits(iota_transaction, trits);
+}
+
+Transaction::~Transaction() { transaction_free(iota_transaction); }
+
+std::string Transaction::get_signature(void) {
+  auto trits = transaction_signature(iota_transaction);
+  tryte_t trytes[2187];
+  flex_trit_to_tryte(trytes, 2187, trits, 6561, 6561);
+  return std::string(reinterpret_cast<char *>(trytes), 2187);
+}
+
+// Set the transaction signature (copy argument)
+void Transaction::set_signature(const std::string &trytes) {
+  size_t flex_len = flex_trits_num_for_trits(6561);
+  flex_trit_t trits[flex_len];
+  tryte_to_flex_trit(trits, 6561, (const tryte_t *)trytes.data(), 2187, 2187);
+  transaction_set_signature(iota_transaction, trits);
+}
+
+// Get the transaction message
+std::string Transaction::get_message(void) {
+  auto trits = transaction_message(iota_transaction);
+  tryte_t trytes[2187];
+  flex_trit_to_tryte(trytes, 2187, trits, 6561, 6561);
+  return std::string(reinterpret_cast<char *>(trytes), 2187);
+}
+
+// Set the transaction message (copy argument)
+void Transaction::set_message(const std::string &trytes) {
+  size_t flex_len = flex_trits_num_for_trits(6561);
+  flex_trit_t trits[flex_len];
+  tryte_to_flex_trit(trits, 6561, (const tryte_t *)trytes.data(), 2187, 2187);
+  transaction_set_message(iota_transaction, trits);
+}
+
+// Get the transaction address
+std::string Transaction::get_address(void) {
+  auto trits = transaction_address(iota_transaction);
+  tryte_t trytes[81];
+  flex_trit_to_tryte(trytes, 81, trits, 243, 243);
+  return std::string(reinterpret_cast<char *>(trytes), 81);
+}
+
+// Set the transaction address (copy argument)
+void Transaction::set_address(const std::string &trytes) {
+  size_t flex_len = flex_trits_num_for_trits(243);
+  flex_trit_t trits[flex_len];
+  tryte_to_flex_trit(trits, 243, (const tryte_t *)trytes.data(), 81, 81);
+  transaction_set_address(iota_transaction, trits);
+}
+
+// Get the transaction value
+int64_t Transaction::get_value(void) {
+  return transaction_value(iota_transaction);
+}
+
+// Set the transaction value
+void Transaction::set_value(int64_t value) {
+  transaction_set_value(iota_transaction, value);
+}
+
+// Get the transaction obsolete tag
+std::string Transaction::get_obsolete_tag(void) {
+  auto trits = transaction_obsolete_tag(iota_transaction);
+  tryte_t trytes[81];
+  flex_trit_to_tryte(trytes, 81, trits, 243, 243);
```

Comment:
as `tryte_t` == `char`, we can avoid the copy by accessing the string's `data()` array directly: https://en.cppreference.com/w/cpp/string/basic_string/data